### PR TITLE
[FLINK-28501] Add Transformer and Estimator for VectorIndexer

### PR DESCRIPTION
**Pull request by zhipeng93**

### CODE REVIEW

File: flink-ml-core/src/main/java/org/apache/flink/ml/linalg/SparseVector.java [hunk 1]
```diff
@@ -56,6 +56,16 @@ public double get(int i) {
         return 0.;
     }
 
+    @Override
+    public void set(int i, double value) {
+        int pos = Arrays.binarySearch(indices, i);
+        if (pos < 0) {
+            throw new IllegalStateException(
+                    "Cannot set value for a SparseVector that does not contain value at that dimension.");
```
[the remark this answers is not in the document]
Comment (zhipeng93) [reply flink-ml-core/src/main/java/org/apache/flink/ml/linalg/SparseVector.java:64]: Good catch. I agree that as long as `i is in [0,n)`, the `set()` should succeed. I have updated the code in the lastest commit.

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/vectorindexer/VectorIndexer.java [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+        DataStream<HashSet<Double>[]> distinctDoubles =
+                DataStreamUtils.reduce(
+                        localDistinctDoubles,
+                        (ReduceFunction<HashSet<Double>[]>)
+                                (value1, value2) -> {
+                                    for (int i = 0; i < value1.length; i++) {
+                                        if (value1[i] == null || value2[i] == null) {
+                                            value1[i] = null;
+                                        } else {
+                                            value1[i].addAll(value2[i]);
```
[the remark this answers is not in the document]
Comment (zhipeng93) [reply flink-ml-lib/src/main/java/org/apache/flink/ml/feature/vectorindexer/VectorIndexer.java:110]: It is actually an implementation issue --- I reused the implementation of `DataStreamUtils#reduce` and there no way to pass the `maxCategories` to the reduce function.

I agree there could be a nicer implementation, but the performance should be similar --- the operation happens on a parallelism-1 task.

**Comment by zhipeng93** [comment]
@yunfengzhou-hub  Thanks for the review. I have updated the PR according to your comments.

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/vectorindexer/VectorIndexer.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (yunfengzhou-hub) [reply flink-ml-lib/src/main/java/org/apache/flink/ml/feature/vectorindexer/VectorIndexer.java:110]: Got it. I agree with it that this change would not bring obvious performance change and thus it is optional.

**Comment by yunfengzhou-hub** [comment]
@zhipeng93 Thanks for the update. It LGTM.

**Pull request merged**